quantization/optimal_quantization.py

```diff
@@ -0,0 +1,171 @@
+"""
+Optimal quantization of the empirical distribution of a dataset -
+
+The algorithm finds an optimal aproximation of the dataset by a
+smaller number n of points, and yields a K-means type clustering.
+The n points of the approximation, weighted by the masses of their
+Voronoi cells, define a discrete distribution that best approximates,
+in the sense of the Wasserstein distance and at fixed support size n,
+the empirical distribution of the dataset. The Voronoi diagram associated
+to these points yields a K-means type clustering of the dataset.
+"""
+
+import numpy as np
+from geomstats.hypersphere import HypersphereMetric
+
+
+TOLERANCE = 1e-5
+IMPLEMENTED = ('S1', 'S2')
+
+
+def sample_from(points, size=1):
+    """
+    Sample from the empirical distribution associated to points.
+    """
+    n_points = points.shape[0]
+    dimension = points.shape[-1]
+
+    index = np.random.randint(low=0, high=n_points, size=size)
+    sample = points[np.ix_(index, np.arange(dimension))]
+
+    return sample
+
+
+def closest_neighbor(point, neighbors, space=None):
+    """
+    Find closest neighbor of point among neighbors.
+    """
+    if space not in IMPLEMENTED:
+        raise NotImplementedError(
+                'The closest neighbor function is not implemented'
+                ' for space {}. The spaces available'
+                ' are: {}.'.format(space, IMPLEMENTED))
+    elif space == 'S1':
+        metric = HypersphereMetric(dimension=1)
+    else:
+        metric = HypersphereMetric(dimension=2)
+
+    dist = metric.dist(point, neighbors)
+    index_closest_neighbor = dist.argmin()
+
+    return index_closest_neighbor
+
+
+def diameter_of_data(points, space=None):
+    """
+    Compute the two points that are farthest away from each other in points.
+    """
+    if space not in IMPLEMENTED:
+        raise NotImplementedError(
+                'The diameter function is not implemented'
+                ' for space {}. The spaces available'
+                ' are: {}.'.format(space, IMPLEMENTED))
+    elif space == 'S1':
+        metric = HypersphereMetric(dimension=1)
+    else:
+        metric = HypersphereMetric(dimension=2)
+
+    diameter = 0.0
+    n_points = points.shape[0]
+
+    for i in range(n_points-1):
+        dist_to_neighbors = metric.dist(points[i, :], points[i+1:, :])
+        dist_to_farthest_neighbor = dist_to_neighbors.max()
+        diameter = np.maximum(diameter, dist_to_farthest_neighbor * 2)
+
+    return diameter
+
+
+def karcher_flow(points, space=None, tolerance=TOLERANCE):
```

Comment:
This looks like: https://github.com/geomstats/geomstats/blob/master/geomstats/riemannian_metric.py#L219

Could we fuse the codes maybe?